### PR TITLE
Xml migration

### PR DESCRIPTION
**Pull request by jonathanfranks**
### Problem/Motivation
XML migrations weren't working. We want a start-to-finish migration flow.

### Proposed solution
This branch adds plugins for forms for data parser and data fetcher configuration, and adds support for unique/ID fields in the migration.

This is the first version of a fully functional built migration, and there are things in here we'll be refactoring and making more generic/universal, but this is the first start of having a working start-to-finish migration workflow.

### Example/QA
__Steps to test:__
- [ ] Go to Structure > Migrations (/admin/structure/migrate)
    - [ ] Press "Add a migration group"  (/admin/structure/migrate/add)
    - [ ] Enter Label, Description, Source type and press "Create Migration Group"
- [ ] Click "List migrations" for the group you just created.
- [ ] Press "Add migration" (/admin/structure/migrate/sources/add)
   - [ ] Add a Label. MAKE SURE YOU SELECT THE MIGRATION GROUP (!!!!!!!!)
   - Use these settings:
     - [ ] Source - Source plugin: url
     - [ ] Data_fetcher - Data_fetcher plugin: http; File Location (single location only): an xml file that Drupal can see
     - [ ] Data_parser - Data_parser plugin: xml; XML Item Selector: your item selector (Irina's sample file is /root/row)
     - [ ] Destination - Destination plugin: entity:node; Content type: article
   - [ ] Press Save.
- [ ] Edit the migration you just created and click the Mapping tab.
- [ ] Press "Add mapping"
   - [ ] Select "Title" for your destination field.
   - [ ] Enter "title" in source. (Again, for Irina's sample XML file)
   - [ ] Check the "Unique Field" box.
   - [ ] Press "Save"
- [ ] Go to Content > Feeds Migrate (/admin/content/feeds-migrate)
   - [ ] Press "Add Importer"
   - Use these settings:
     - [ ] Label: New feeds importer
     - [ ] Migration source: the migration you just created
     - [ ] Settings - import period: off
     - [ ] Processor Settings - Update existing content: Do not update existing content; Orphaned items: keep
     - [ ] Data Fetcher Settings - URL: the same URL you put in the data fetcher setting in the migration
   - [ ] Press "Save"
Click "Import" on the importer you just created.
It imports your content!

### CODE REVIEW

File: src/MappingFieldFormBase.php [hunk 1]
```diff
@@ -120,6 +120,13 @@ public function buildConfigurationForm(array $form, FormStateInterface $form_sta
       '#default_value' => $mapping['source'],
     ];
 
+    $checked = array_key_exists($mapping['source'], $this->migration->source["ids"]);
```
Comment (Etroid) [inline src/MappingFieldFormBase.php:123]: Nit, maybe `$unique` as a variable name is more descriptive

File: modules/feeds_migrate_ui/src/Form/MigrationMappingForm.php [hunk 2]
```diff
@@ -298,11 +298,12 @@ protected function buildTableRow(array $mapping, $property = NULL) {
 
     // Unique.
     if (!$property) {
+      $checked = array_key_exists($source, $migration->source['ids']);
```
Comment (Etroid) [inline modules/feeds_migrate_ui/src/Form/MigrationMappingForm.php]: Nit, maybe more descriptive `$unique` variable

File: modules/feeds_migrate_ui/src/Form/MigrationMappingFormBase.php [hunk 3]
```diff
@@ -322,11 +332,59 @@ public function submitForm(array &$form, FormStateInterface $form_state) {
...
+
+    // Add the fields to the source section.
+    //    fields:
+    //      -
+    //      name: title
+    //      label: Title
+    //      selector: title
+
+    $fields = $source['fields'] ?: [];
+    $foundField = FALSE;
```
Comment (Etroid) [inline modules/feeds_migrate_ui/src/Form/MigrationMappingFormBase.php:370]: Nit, we should standardize on `camelCase` versus `snake_case` for local variables at some point. 🤷‍♂️

File: src/Plugin/feeds_migrate/data_fetcher/File.php [hunk 4]
```diff
@@ -83,7 +83,7 @@ public function alterMigration(FeedsMigrateImporterInterface $importer, Migratio
       if ($file = FileEntity::load(reset($fids['file']))) {
 
         $source_config = $migration->getSourceConfiguration();
-        $source_config['urls'] = file_create_url($file->getFileUri());
+        $source_config['urls'] = \Drupal::service('file_system')->realpath($file->getFileUri());
```
Comment (Etroid) [inline src/Plugin/feeds_migrate/data_fetcher/File.php]: Maybe add a todo to use DI for this service?

**Review by Etroid — APPROVED** [review]
This is great! Minor comments, but nothing blocking. 🏆

File: modules/feeds_migrate_ui/src/Form/MigrationMappingFormBase.php (hunk 3 above)
Comment (jonathanfranks) [reply modules/feeds_migrate_ui/src/Form/MigrationMappingFormBase.php:370]: I do `camelCase` in class and `snake_case` in a module file. I think that's the standard, right? :)

Comment (Etroid) [reply modules/feeds_migrate_ui/src/Form/MigrationMappingFormBase.php:370]: Sounds good!

File: src/Plugin/feeds_migrate/data_fetcher/File.php (hunk 4 above)
Comment (jonathanfranks) [reply src/Plugin/feeds_migrate/data_fetcher/File.php]: Added DI.

File: modules/feeds_migrate_ui/src/Form/MigrationMappingForm.php (hunk 2 above)
Comment (jonathanfranks) [reply modules/feeds_migrate_ui/src/Form/MigrationMappingForm.php]: Done.

**Pull request merged**